File: src/main/java/com/github/shatteredsuite/scrolls/items/VersionConverter.java
package com.github.shatteredsuite.scrolls.items;

import com.github.shatteredsuite.core.include.nbt.NBTItem;
import com.github.shatteredsuite.scrolls.ShatteredScrolls2;
import com.github.shatteredsuite.scrolls.data.scroll.BindingData;
import com.github.shatteredsuite.scrolls.data.scroll.LocationBindingData;
import com.github.shatteredsuite.scrolls.data.scroll.ScrollType;
import com.github.shatteredsuite.scrolls.data.scroll.UnboundBindingData;
import com.github.shatteredsuite.scrolls.data.scroll.WarpBindingData;
import org.bukkit.Location;
import org.bukkit.inventory.ItemStack;

public class VersionConverter {
    public static ScrollInstance createScrollInstanceFromV1(ItemStack stack) {
        ScrollType newType = ShatteredScrolls2.getInstance().config().defaultType;
        NBTItem item = new NBTItem(stack);
        if(!item.hasKey("shatteredscrolls_bound")) {
            return null;
        }
        int charges = item.getInteger("shatteredscrolls_charges");
        boolean infinite = charges == -1024;
        BindingData bindingData;
        if(item.getByte("shatteredscrolls_bound") == 1) {
            if(item.hasKey("shatteredscrolls_destination")) {
                bindingData = new WarpBindingData(ShatteredScrolls2.getInstance().warps()
                    .get(item.getString("shatteredscrools_destination")));
            }
            else {
                Location location = NBTUtils.locationFromNBTItem(item, "shatteredscrolls_dest_");
                bindingData = new LocationBindingData(location);
            }
        } else {
            bindingData = new UnboundBindingData();
        }
        return new ScrollInstance(newType, charges, infinite, bindingData);
    }
}

File: src/main/kotlin/com/github/shatteredsuite/scrolls/items/VersionConverter.kt
package com.github.shatteredsuite.scrolls.items

import com.github.shatteredsuite.core.include.nbt.NBTItem
import com.github.shatteredsuite.scrolls.ShatteredScrolls
import com.github.shatteredsuite.scrolls.data.scroll.binding.BindingData
import com.github.shatteredsuite.scrolls.data.scroll.binding.LocationBindingData
import com.github.shatteredsuite.scrolls.data.scroll.binding.UnboundBindingData
import com.github.shatteredsuite.scrolls.data.scroll.binding.WarpBindingData
import org.bukkit.inventory.ItemStack

object VersionConverter {
    @JvmStatic
    fun createScrollInstanceFromV1(stack: ItemStack): ScrollInstance? {
        val inst = ShatteredScrolls.getInstance()
        val newType = inst.config().getDefaultType()
        val item = NBTItem(stack)
        if (!item.hasKey("shatteredscrolls_bound")) {
            return null
        }
        val charges = item.getInteger("shatteredscrolls_charges")
        val infinite = charges == -1024
        val bindingData: BindingData
        bindingData = if (item.getByte("shatteredscrolls_bound").toInt() == 1) {
            if (item.hasKey("shatteredscrolls_destination")) {
                WarpBindingData(inst.warps()[item.getString("shatteredscrolls_destination")])
            } else {
                val location = NBTUtils.locationFromNBTItem(item, "shatteredscrolls_dest_")
                LocationBindingData(location)
            }
        } else {
            UnboundBindingData()
        }
        return ScrollInstance(newType, charges, infinite, bindingData)
    }
}